### PR TITLE
model: support NVILA and NVILA Lite

### PR DESCRIPTION
**Pull request by futrime**
## Motivation

Add support for NVILA and NVILA Lite.

## Modifications

1. Add NVILA and NVILA Lite model and processor impl.
2. Split NVILA and NVILA Lite into two standalone model impl to match their hf impl.
3. Add `block_sizes` to multimodal items for NVILA.
4. Fix type annotation for BaseMultimodalProcessor.load_mm_data().
5. Add tests for NVILA-Lite-2B and NVILA-8B.

## Accuracy Tests

| Model | MMMU / hf | MMMU / SGLang |
| -- | -- | -- |
| NVILA-Lite-2B |0.373 | 0.372 |
| NVILA-8B | 0.478 | 0.472 |

## Benchmarking and Profiling

N/A.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/multimodal/processors/nvila.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            ),
+            video_data=(
+                request_obj.video_data
+                if isinstance(request_obj.video_data, list)
+                else [request_obj.video_data]
+            ),
+        )
+
+        for i, video in enumerate(base_output.videos):  # type: ignore
+            base_output.videos[i] = [x.asnumpy() for x in video]  # type: ignore
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/multimodal/processors/nvila.py:63]: base_output.videos seems to be decord VideoReader.

File: python/sglang/srt/multimodal/processors/nvila_lite.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            ),
+            video_data=(
+                request_obj.video_data
+                if isinstance(request_obj.video_data, list)
+                else [request_obj.video_data]
+            ),
+        )
+
+        for i, video in enumerate(base_output.videos):  # type: ignore
+            base_output.videos[i] = [x.asnumpy() for x in video]  # type: ignore
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/multimodal/processors/nvila_lite.py]: base_output.videos seems to be decord VideoReader.

File: python/sglang/srt/models/nvila.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    def __init__(self, config: NVILAConfig):
+        super().__init__()
+
+        self.layers = nn.Sequential(
+            NVILAMultiModalProjectorDownsampleBlock(),
+            nn.LayerNorm(MM_HIDDEN_SIZE * 4),
+            nn.Linear(MM_HIDDEN_SIZE * 4, config.text_config.hidden_size),
+            nn.GELU(),
+            nn.Linear(config.text_config.hidden_size, config.text_config.hidden_size),
+        )
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/models/nvila.py:97]: MM_HIDDEN_SIZE does not equal to vision tower's hidden size.

File: python/sglang/srt/models/nvila.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+    def get_image_feature(self, mm_input: list[MultimodalDataItem]) -> Tensor:
+        block_sizes = (
+            list(
+                itertools.chain.from_iterable(
+                    x.block_sizes for x in mm_input if hasattr(x, "block_sizes")
+                )
+            )
+            or None
+        )
+        pixel_values = torch.cat([torch.tensor(x.feature) for x in mm_input], dim=0)
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/models/nvila.py:154]: x.feature might be ndarray.

File: python/sglang/srt/models/nvila_lite.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        self.layers = nn.Sequential(
+            NVILALiteMultiModalProjectorDownsampleBlock(),
+            nn.LayerNorm(MM_HIDDEN_SIZE * 9),
+            nn.Linear(MM_HIDDEN_SIZE * 9, MM_HIDDEN_SIZE * 3),
+            nn.GELU(),
+            nn.LayerNorm(MM_HIDDEN_SIZE * 3),
+            nn.Linear(MM_HIDDEN_SIZE * 3, config.text_config.hidden_size),
+            nn.GELU(),
+            nn.Linear(config.text_config.hidden_size, config.text_config.hidden_size),
+        )
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/models/nvila_lite.py:99]: same.

File: python/sglang/srt/models/nvila_lite.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+            get_embedding=get_embedding,
+            positions=positions,
+        )
+
+        assert isinstance(output, LogitsProcessorOutput)
+
+        return output
+
+    def get_image_feature(self, mm_input: list[MultimodalDataItem]) -> Tensor:
+        pixel_values = torch.cat([torch.tensor(x.feature) for x in mm_input], dim=0)
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/models/nvila_lite.py:148]: might be ndarray.

File: python/sglang/srt/models/nvila.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+        selected_layer_hidden_states = vision_tower_output.hidden_states[
+            self.config.mm_vision_select_layer
+        ]
+
+        if self.config.mm_vision_select_feature == "cls_patch":
+            return selected_layer_hidden_states
+        else:
+            raise NotImplementedError(
+                f"Unsupported mm_vision_select_feature: {self.config.mm_vision_select_feature}"
+            )
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/models/nvila.py]: ok i'll remove it soon.

File: python/sglang/srt/models/nvila_lite.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        selected_layer_hidden_states = vision_tower_output.hidden_states[
+            self.config.mm_vision_select_layer
+        ]
+
+        if self.config.mm_vision_select_feature == "cls_patch":
+            return selected_layer_hidden_states
+        else:
+            raise NotImplementedError(
+                f"Unsupported mm_vision_select_feature: {self.config.mm_vision_select_feature}"
+            )
```
[the remark this answers is not in the document]
Comment (futrime) [reply python/sglang/srt/models/nvila_lite.py]: ok i'll remove it soon.

File: python/sglang/srt/models/nvila.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+from sglang.srt.managers.mm_utils import MultiModalityDataPaddingPatternMultimodalTokens
+from sglang.srt.managers.schedule_batch import (
+    Modality,
+    MultimodalDataItem,
+    MultimodalInputs,
+)
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+from sglang.srt.models.qwen2 import Qwen2ForCausalLM
+
+MM_HIDDEN_SIZE = 3456
```
Comment (mickqian) [inline python/sglang/srt/models/nvila.py:30]: why is this hard-coded?

File: python/sglang/srt/multimodal/processors/nvila.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            multimodal_tokens=self.mm_tokens,
+            image_data=(
+                request_obj.image_data
+                if isinstance(request_obj.image_data, list)
+                else [request_obj.image_data]
+            ),
+            video_data=(
+                request_obj.video_data
+                if isinstance(request_obj.video_data, list)
+                else [request_obj.video_data]
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/nvila.py]: If I remember correctly, this is always a list?

File: python/sglang/srt/multimodal/processors/nvila_lite.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+        mm_items, input_ids, _ = self.process_and_combine_mm_data(
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/nvila_lite.py]: Can we share these two processors?

**Comment by Lyken17** [comment]
@futrime  one more comment, can you also add NVILA series to the doc page https://docs.sglang.ai/supported_models/multimodal_language_models.html here?

**Comment by futrime** [comment]
> @futrime one more comment, can you also add NVILA series to the doc page https://docs.sglang.ai/supported_models/multimodal_language_models.html here?

https://github.com/sgl-project/sglang/pull/10399/files#diff-4ef7040d6912fabeaef33c9f2d2c61166c8e3b99fe1d7611343a83daad53a8d5

File: python/sglang/srt/models/nvila.py (hunk 9 above)
Comment (futrime) [reply python/sglang/srt/models/nvila.py:30]: mm hidden size is not present in NVILA's config

Comment (adarshxs) [inline python/sglang/srt/models/nvila.py:30]: Can you pull this value from the config?
https://huggingface.co/Efficient-Large-Model/NVILA-8B/blob/main/config.json#L100

File: python/sglang/srt/models/nvila_lite.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+from sglang.srt.managers.mm_utils import MultiModalityDataPaddingPatternMultimodalTokens
+from sglang.srt.managers.schedule_batch import (
+    Modality,
+    MultimodalDataItem,
+    MultimodalInputs,
+)
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+from sglang.srt.models.qwen2 import Qwen2ForCausalLM
+
+MM_HIDDEN_SIZE = 1152
```
Comment (adarshxs) [inline python/sglang/srt/models/nvila_lite.py:29]: this too: https://huggingface.co/Efficient-Large-Model/NVILA-Lite-8B/blob/main/config.json#L100

File: python/sglang/srt/models/nvila.py (hunk 9 above)
Comment (futrime) [reply python/sglang/srt/models/nvila.py:30]: The HuggingFace-compatible checkpoint of NVILA lies at https://huggingface.co/Efficient-Large-Model/NVILA-8B-hf/blob/main/config.json and it doesn't contain a `mm_hidden_size`.

File: python/sglang/srt/models/nvila_lite.py (hunk 12 above)
Comment (futrime) [reply python/sglang/srt/models/nvila_lite.py:29]: Same as above

File: python/sglang/srt/models/nvila.py (hunk 9 above)
Comment (adarshxs) [reply python/sglang/srt/models/nvila.py:30]: Ok

**Comment by fy1214** [comment]
thanks @2015aroras. Could you provide the benchmarking latency? like bench one batch to show the throughtput.

File: python/sglang/srt/models/nvila.py [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+                    F.interpolate(
+                        cur_features_each_scale[i].to(torch.float32),
+                        size=output_size,
+                        mode="area",
+                    ).to(cur_features_each_scale[i].dtype)
+                    for i in range(len(cur_features_each_scale))
+                ],
+                dim=1,
+            )
+            # cur_features = rearrange(cur_features, "1 c h w -> (h w) c")
```
Comment (mickqian) [inline python/sglang/srt/models/nvila.py]: remove this?

File: test/srt/test_vision_openai_server_b.py [hunk 14]
```diff
@@ -241,6 +241,42 @@ def setUpClass(cls):
         cls.base_url += "/v1"
 
 
+class TestNVILAServer(ImageOpenAITestMixin, VideoOpenAITestMixin):
```
Comment (mickqian) [inline test/srt/test_vision_openai_server_b.py]: Please add it to nightly-ci instead, refer to `sglang/test/README.md`

**Review by mickqian — CHANGES_REQUESTED** [review]

**Comment by mickqian** [comment]
Also, please rebase

**Comment by futrime** [comment]
```
Prefill. latency: 0.02420 s, throughput:  42308.96 token/s
Decode 0. Batch size: 1, latency: 0.00416 s, throughput:    240.31 token/s
Decode 1. Batch size: 1, latency: 0.00414 s, throughput:    241.66 token/s
Decode 2. Batch size: 1, latency: 0.00426 s, throughput:    234.91 token/s
Decode 3. Batch size: 1, latency: 0.00416 s, throughput:    240.40 token/s
Decode 4. Batch size: 1, latency: 0.00414 s, throughput:    241.72 token/s
Decode.  median latency: 0.00413 s, median throughput:    242.24 token/s
Total. latency:  0.086 s, throughput:  12049.99 token/s
```

**Review by fy1214 — APPROVED** [review]
1、no other main code
2、benchmark accuracy and latency
3、update the sglang doc

**Review by zhaochenyang20 — APPROVED** [review]
Great work. Thanks for @fy1214 for review

**Comment by Lyken17** [comment]
@zhaochenyang20 only 1 test failed seems due to CI error. Can you help re-run and merge?
